Have `pip cache purge` and `pip cache remove` handle empty directories.

duckinator opened the pull request:
<!---
Thank you for your soon to be pull request. Before you submit this, please
double check to make sure that you've added a news file fragment. In pip we
generate our NEWS.rst from multiple news fragment files, and all pull requests
require either a news file fragment or a marker to indicate they don't require
one.

To read more about adding a news file fragment for your PR, please check out
our documentation at: https://pip.pypa.io/en/latest/development/contributing/#news-entries
-->

These are the changes discussed in #7372:

1. [x] Make `pip cache remove` prune empty directories.
2. [x] Make `pip cache purge` delete everything in the http and wheels caches (not just files).
3. [x] Make `pip cache purge` remove `selfchecks.json`.
4. [ ] Make `pip cache purge` remove things in the `pip/selfchecks/` directory when appropriate.

Anything not checked off above is not handled by this PR.

pradyunsg commented:
Heads up: The MacOS workers on Azure Pipelines are failing right now -- #9030.

duckinator commented:
Thanks for the heads up. I also just discovered that `os.scandir()` didn't arrive until Python 3.5, so will need to reimplement that in terms of `os.walk()` real quick.

pradyunsg commented:
You could wait for a couple of weeks, and we'll drop Python 2 support from `master`. :)

duckinator commented:
Making it work without using `os.scandir()` is apparently pretty annoying.

Do y'all support Python 3 versions before 3.5? If you don't then, screw it, I'll just wait for you to drop Python 2 support lmao

pradyunsg commented:
> Do y'all support Python 3 versions before 3.5?

Nope. Python 3.5 and Python 2 get dropped after 20.3. :)

duckinator commented:
Okay, in that case let's leave this for after Py2 is dropped. :+1:

duckinator commented:
Rebased off of master (ab7ff0a).

duckinator commented:
Rebased off master (d108e49).

I'm not sure what the _reason_ for the failing tests is. I do see this error in the failed tasks, however:

> ERROR: InvocationError for command /Users/runner/work/pip/pip/.tox/lint/bin/pre-commit run --all-files --show-diff-on-failure --hook-stage=manual (exited with code 1)

duckinator commented:
As a note, anything requiring more than rebasing this PR will probably have to wait a month or two at least.

uranusjr commented:
This seems to be the specific error:

```
flake8...................................................................Failed
- hook id: flake8
- exit code: 1

src/pip/_internal/commands/cache.py:190:17: E127 continuation line over-indented for visual indent
```

duckinator commented:
ah, thank you @uranusjr. Had trouble finding that in the CI output. :sweat_smile:  

I fixed that problem (or at least, that's what `tox -e lint` says locally), but it still ends with this and has exit code 1, for some reason:

```
isort....................................................................Passed
mypy.....................................................................Passed                            
use logger.warning(......................................................Passed
check for eval().........................................................Passed
rst ``code`` is two backticks............................................Passed
NEWS fragment........................................(no files to check)Skipped
check-manifest...........................................................Passed
ERROR: InvocationError for command /home/puppy/dev/python/duckinator/pip/.tox/lint/bin/pre-commit run --all-files --sho
w-diff-on-failure --hook-stage=manual (exited with code 1)
_______________________________________________________ summary _______________________________________________________
ERROR:   lint: commands failed
```

Not sure what's up with that.

sbidoul commented:
@duckinator the flake8 error seems to be [still present](https://github.com/pypa/pip/runs/1824540187#step:8:29).

BrownTruck commented:
Hello!

I am an automated bot and I have noticed that this pull request is not currently able to be merged. If you are able to either merge the ``master`` branch into this pull request or rebase this pull request against ``master`` then it will be eligible for code review and hopefully merging!

pradyunsg commented:
Ahoy @duckinator -- would you be interested in updating this PR? :)

duckinator commented:
@pradyunsg I'll rebase this sometime in the next few days. If it's not done by Wednesday (September 22nd) feel free to ping me here again. 👍

duckinator commented:
Running a bit behind on things this week, but still hoping to get to it in the next few days. Sorry for the delay!

pradyunsg commented:
No worries and thanks for the update! Please don’t feel pressured to update this ASAP. :)

I’m gonna put this into the release milestone to remind myself to keep an eye out for updates here — it’s totally fine if this gets pushed down the road.

duckinator commented:
Rebased, and simplified(!) because some of the changes in the last ~year let me simplify my code. ^.^

pradyunsg commented on src/pip/_internal/commands/cache.py:
```diff
@@ -180,7 +180,21 @@ def remove_cache_items(self, options: Values, args: List[Any]) -> None:
+            self._cache_dir(options, "http")
+        ) + filesystem.list_empty_subdirs(self._cache_dir(options, "wheels"))
+        for dirname in dirs:
+            os.rmdir(dirname)
```
Should we have a verbose log here?

duckinator replied:
Probably, and done!

pradyunsg commented on src/pip/_internal/utils/filesystem.py:
```diff
@@ -180,3 +180,12 @@ def directory_size(path: str) -> Union[int, float]:
+    """Returns a list of absolute paths of empty directories beneath path."""
+    result = []  # type: List[str]
+    for root, dirs, _files in os.walk(path, topdown=False):
+        result.extend(os.path.join(root, d) for d in dirs)
```
We can `yield from ...` here and avoid creating a list.

duckinator replied:
In the process of trying to figure out how to use `yield from ...` here, I found a bug in `list_empty_subdirs()`. The bug is it doesn't even check if the directory is empty. :joy:

pradyunsg replied:
🤣

duckinator commented:
Pulling off the discussion of "whoops it never actually checked if directories are empty" into the main thread because that was an unexpectedly fascinating can of worms.

Saying the function was looking for "empty" directories was actually inaccurate. It was looking for directories where there are no files in _that directory or its subdirectories_, but having directories in it is fine.

Apparently this is **_difficult_**.

The way I wound up doing it is finding "leaf" directories (directories with no files OR directories in them) then also adding all the parent directories until files were found or a fixed point was hit.

This isn't the best-performing approach, and can likely be improved, but it's the only one I could really wrap my head around at the moment.

duckinator commented:
@pradyunsg assuming all the tests (except `Check / news entry`) pass, this is good for a review. I'm a little uncomfortable with how much code is needed for `utils.filesystems.subdirs_with_no_files(path)`. However, realistically, that _is_ a fairly complex problem, so I don't know if the code can be made much simpler than it is now.

duckinator commented:
>.> I forgot to run `tox -e lint` on it. Whoops.

Should be good _now_.

pradyunsg commented:
Ooops, our codebase is fully typed now, and this PR adds functions without types. 😅

sbidoul commented on src/pip/_internal/utils/filesystem.py:
```diff
@@ -180,3 +182,50 @@ def directory_size(path: str) -> Union[int, float]:
+def subdirs_with_no_files(path):
+    """Yields every subdirectory of +path+ that has no files under it."""
+    for leaf in _leaf_subdirs(path):
+        yield from _leaf_parents_without_files(path, leaf)
```
Do we want a couple of unit tests for this function ?

duckinator replied:
Yeah, probably.

pradyunsg commented:
To communicate with sufficient clarity: I do not intend to block the 21.3 release on this PR, since this doesn't seem to be a show-stopper issue; even though this would be a nice-to-have improvement!

To that end, I've gone ahead and dropped this from the release milestone. :)

duckinator commented:
TODO:

* [ ] Add type information.
* [ ] Add tests for `subdirs_with_no_files(path)`.

duckinator commented:
Hi, sorry for disappearing for a while there. Hoping to get back to this sometime this month, so it can be properly wrapped up and merged.

uranusjr commented:
Gentle ping

duckinator commented:
@uranusjr hey, thanks for the nudge. Got a bit overwhelmed with other things, and this fell through the cracks.

duckinator commented:
Added type information to all the functions I added in `src/pip/_internal/utils/filesystem.py`. Last remaining thing is tests for `pip._internals.utils.filesystem.subdirs_with_no_files(path_str)`.

uranusjr commented on src/pip/_internal/commands/cache.py:
```diff
@@ -184,7 +184,24 @@ def remove_cache_items(self, options: Values, args: List[Any]) -> None:
+        wheel_dirs = filesystem.subdirs_with_no_files(
+            self._cache_dir(options, "wheels")
+        )
+        dirs = list(http_dirs) + list(wheel_dirs)
+        for dirname in dirs:
```
Use `itertools.chain` here

duckinator replied:
as in something like this?:

```python
dirs = itertools.chain(http_dirs, wheel_dirs)
for dirname in dirs:
    ...
```

(Note that `dirs` is used on line 204, so I can't get away with `for dirname in itertools.chain(...)` as easily.)

uranusjr replied:
Not exactly because you can’t use `len` on an iterator.

Something like this

```python
for dirname in itertools.chain(http_dirs, wheel_dirs):
    ...

...

logger.info("Empty directories removed: %s", len(http_dirs) + len(wheel_dirs))
```

duckinator replied:
Done. 👍

duckinator replied:
So, turns out `http_dirs` and `wheel_dirs` are defined with results from `filesystem.subdirs_with_no_files(path_str)` which returns a Generator, so `len(http_dirs)` and `len(wheel_dirs)` won't work:

```
src/pip/_internal/commands/cache.py:204: error: Argument 1 to "len" has
incompatible type "Generator[str, None, None]"; expected "Sized"  [arg-type]
    ...logger.info("Empty directories removed: %s", len(http_dirs) + len(whee...
                                                        ^
```

Serves me right for not having a local dev environment and just assuming it'd work because it looked right. 😂

duckinator commented:
Included @uranusjr's feedback and rebased off c247ddc.

Still need to add tests for `subdirs_with_no_files(path_str)`.

sbidoul commented:
To implement https://github.com/pypa/pip/issues/5037, I am going to add some metadata files in the wheel cache entries directories. So instead of checking for empty directories, we may want to check for directories with no `.whl` remaining files.

duckinator commented:
@sbidoul thanks for the heads-up. I added an updated TODO in the first comment (so it's not buried), which includes accounting for that. 👍

duckinator commented:
Rebased off main (https://github.com/pypa/pip/commit/dc7abcada7e263e80af6c8d4ea2147c067c78d01).

duckinator commented:
Rebased off main (81f6a9fcc6eb846c58044de61c04b95e90b96a7a).

uranusjr commented:
Need to update tests to match the new output.

sbidoul commented:
Since 22.2 introduced `origin.json` in cache entries, and with https://github.com/pypa/pip/pull/11897, this PR and it's first remaining TODO becomes more important:

> Check for directories without .whl files, instead of empty directories

duckinator commented:
Rebased off main (3682309) and updated tests to match the new output format.

uranusjr commented on src/pip/_internal/commands/cache.py:
```diff
@@ -183,7 +183,28 @@ def remove_cache_items(self, options: Values, args: List[Any]) -> None:
+        wheel_dirs = filesystem.subdirs_with_no_files(
+            self._cache_dir(options, "wheels")
+        )
+        dirs = list(http_dirs) + list(wheel_dirs)
```
```suggestion
        dirs = [*http_dirs, *wheel_dirs]
```

This should reduce memory usage a bit it avoids building the intermediate lists.

notatallshaw commented:
Can you add a news entry please: https://pip.pypa.io/en/latest/development/contributing/#news-entries

ichard26 commented:
@duckinator glad to see that you aren't (too?) frustrated that we left this PR out to dry so long it probably fossilized. Just as a FYI, there is (surprise!) still limited review capacity so it may take some time for us to review your PR. However, I do want to see your work be merged at some point, so when I come back from my OSS "break", I'll make sure to review your PR!

duckinator commented:
@ichard26 honestly, I was worried I was annoying y'all because I kept disappearing for _months_ or _years_ at a time then coming back and doing a barrage of work only to disappear again. :sweat_smile: 

Looking forward to getting this wrapped up whenever you get to it. Enjoy your break. :slightly_smiling_face:

duckinator commented on src/pip/_internal/commands/cache.py:
```diff
@@ -189,7 +189,31 @@ def remove_cache_items(self, options: Values, args: list[str]) -> None:
+                for file in subdir.iterdir():
+                    file.unlink(missing_ok=True)
+                subdir.rmdir()
+            except FileNotFoundError:
+                # If the file is already gone, that's fine.
+                pass
```
Yes, that's a typo. It's the `rmdir()` call that can raise `FileNotFoundError`.

duckinator commented:
Thanks for the review! I've responded to the things you mentioned (& accepted your commit suggestion).

I'll try to work on this over the next few days. Please feel free to ping me if you don't hear anything before August 17th &mdash; I don't want it to slip through the cracks again! :slightly_smiling_face: 

> I finally got around to doing a review. The idea looks good, but I think the implementation can be simplified. Sorry for making you do more work :)

No worries! I could tell it was rough, and expected it to need a second pass.

> Also, where are the unit tests? The discussion history seems to suggest that they were written at some point. Did they forgotten to be committed?

I'm not sure what happened here. If I did write them, I suspect they got lost when I switched OSes at some point in the last 5 years.

> Thanks a lot for your persistence!

Glad to help! I'm really hoping I can get it wrapped up before the half-decade mark rolls around in October. :joy:

duckinator commented:
I didn't realize that `Path.walk()` was added in Python 3.12. I switched to `os.walk()`, so it should work on old Python versions.
